### PR TITLE
Add Zarr storage

### PR DESCRIPTION
**Pull request by tomwhite**
This change allows Anndata to use [Zarr](http://zarr.readthedocs.io/) as the backing storage. The motivation here is to provide support for Zarr with [Scanpy](http://scanpy.readthedocs.io/) and Spark (although this patch does not have any Spark parts in it). This work can be found in [this repo](https://github.com/lasersonlab/single-cell-experiments).

The code here is based on the HDF5 codepath, although it might be possible to simplify it somewhat, and it also needs docstrings adding. I wanted to submit the PR now to see if there is general interest and if such a change would be considered.

### CODE REVIEW

**Comment by falexwolf** [comment]
Ah! Cool! Thanks for pointing this out...

We are currently redesigning the backed infrastructure of AnnData anyways, so this comes at the right time.

Looking at the pull request, I see some conflicts. I also see some copy and paste - I'm adding a comment in the code... It would be great to discuss this a little.

@Koncopd your opinion is also valid... how far are you with the new infrastructure? let's put your fork on a development branch here, so that we won't run into further conflicts...

File: anndata/readwrite/read.py [hunk 1]
```diff
@@ -321,6 +321,55 @@ def _read_text(
...
+def read_zarr(store):
+    import zarr
+    f = zarr.open(store, mode='r')
+    d = {}
+    for key in f.keys():
+        _read_key_value_from_zarr(f, d, key)
+    return AnnData(d)
+
+
+def _read_key_value_from_zarr(f, d, key, key_write=None):
```
Comment (falexwolf) [inline anndata/readwrite/read.py]: this is very redundant, instead we should have a helper that is called both by the HDF5 functions and the zarr... @Koncopd what do you think?

**Comment by falexwolf** [comment]
If you grant us access to work on your pull request, we can also see if we converge in this way and add the changes that I'd like to see ourselves....

**Comment by tomwhite** [comment]
@falexwolf thanks for the feedback! I agree about eliminating the redundancy with the HDF5 code.

Please go ahead and add any edits you see fit - you should have write access.

**Comment by falexwolf** [comment]
Ok, sorry for the long wait. I'm merging this now... we're updating several aspects of the functionality and will also address the questions that I raised above.

**Pull request merged**